mediadevices.go

```diff
@@ -142,18 +145,55 @@ func selectBestDriver(filter driver.FilterFn, constraints MediaTrackConstraints)
 }
 
 func (m *mediaDevices) selectAudio(constraints MediaTrackConstraints) (Tracker, error) {
-	d, c, err := selectBestDriver(driver.FilterAudioRecorder(), constraints)
+	filter := driver.FilterAudioRecorder()
+	if constraints.DeviceID != "" {
+		filter = func(d driver.Driver) bool {
+			return driver.FilterAudioRecorder()(d) && driver.FilterID(constraints.DeviceID)(d)
+		}
+	}
+
+	d, c, err := selectBestDriver(filter, constraints)
 	if err != nil {
 		return nil, err
 	}
 
 	return newAudioTrack(m.codecs[webrtc.RTPCodecTypeAudio], d, c)
 }
 func (m *mediaDevices) selectVideo(constraints MediaTrackConstraints) (Tracker, error) {
-	d, c, err := selectBestDriver(driver.FilterVideoRecorder(), constraints)
+	filter := driver.FilterVideoRecorder()
+	if constraints.DeviceID != "" {
+		filter = func(d driver.Driver) bool {
+			return driver.FilterVideoRecorder()(d) && driver.FilterID(constraints.DeviceID)(d)
+		}
+	}
+
+	d, c, err := selectBestDriver(filter, constraints)
 	if err != nil {
 		return nil, err
 	}
 
 	return newVideoTrack(m.codecs[webrtc.RTPCodecTypeVideo], d, c)
 }
+
+func (m *mediaDevices) EnumerateDevices() []MediaDeviceInfo {
+	drivers := driver.GetManager().Query(
+		driver.FilterFn(func(driver.Driver) bool { return true }))
+	info := make([]MediaDeviceInfo, 0, len(drivers))
+	for _, d := range drivers {
+		var kind MediaDeviceType
+		switch {
```

Comment:
I like this. It makes sense. Instead of using type inference, we can use these functions to hide that. Pretty cool!